### ISSUE TITLE
Formal testing system

### ISSUE THREAD

**Issue by dkoslicki** [RTXteam/RTX#781]
- [ ] unit tests
- [ ] integration tests
- [ ] in a separate folder

**Comment by dkoslicki** [comment]
Relevant SRI doc [here](https://docs.google.com/document/d/1aR_o_Mvr1cWqKrqVPj_irXPyDcbeGFtVGh9-ac0cARU/edit).

**Comment by edeutsch** [comment]
Testing system is in progress. I think I have something that will work pretty well. Still a bit more work to do before ready to check it in.

**Comment by dkoslicki** [comment]
Current condition of this:
```bash
cd RTX/code/ARAX/test
pytest -v
```

**Comment by dkoslicki** [comment]
@edeutsch as you wanted someone to test this, here's the current output of the above command:
```bash
RTX/code/ARAX/test$ pytest -v
============================ test session starts ============================
platform linux -- Python 3.8.1, pytest-5.4.3, py-1.8.1, pluggy-0.13.1 -- /home/dkoslicki/anaconda3/envs/RTX/bin/python
cachedir: .pytest_cache
rootdir: /home/dkoslicki/Desktop/RTX/code/ARAX/test
collected 12 items                                                          

test_ARAX_messenger.py::test_create_message_basic PASSED              [  8%]
test_ARAX_messenger.py::test_create_message_node_edge_types PASSED    [ 16%]
test_ARAX_messenger.py::test_add_qnode_basic PASSED                   [ 25%]
test_ARAX_messenger.py::test_add_qnode_curie_scalar PASSED            [ 33%]
test_ARAX_messenger.py::test_add_qnode_curie_list PASSED              [ 41%]
test_ARAX_messenger.py::test_add_qnode_name PASSED                    [ 50%]
test_ARAX_messenger.py::test_add_qnode_type PASSED                    [ 58%]
test_ARAX_messenger.py::test_add_qnode_bad_name PASSED                [ 66%]
test_ARAX_messenger.py::test_add_qnode_bad_parameters PASSED          [ 75%]
test_ARAX_messenger.py::test_add_qedge_multitest PASSED               [ 83%]
test_ARAX_query.py::test_query_by_canned_query_Q0 PASSED              [ 91%]
test_ARAX_query.py::test_query_by_query_graph_2 FAILED                [100%]

================================= FAILURES ==================================
________________________ test_query_by_query_graph_2 ________________________

    def test_query_by_query_graph_2():
        query = { "message": { "query_graph": { "edges": [
                    { "id": "qg2", "source_id": "qg1", "target_id": "qg0", "type": "physically_interacts_with" }
                ],
                "nodes": [
                    { "id": "qg0", "name": "acetaminophen", "curie": "CHEMBL.COMPOUND:CHEMBL112", "type": "chemical_substance" },
                    { "id": "qg1", "name": None, "desc": "Generic protein", "curie": None, "type": "protein" }
                ] } } }
        araxq = ARAXQuery()
        result = araxq.query(query)
>       assert result.status == 'OK'
E       AssertionError: assert 'ERROR' == 'OK'
E         - OK
E         + ERROR

test_ARAX_query.py:35: AssertionError
========================== short test summary info ==========================
FAILED test_ARAX_query.py::test_query_by_query_graph_2 - AssertionError: a...
======================= 1 failed, 11 passed in 2.36s ========================
```

**Comment by edeutsch** [comment]
great, thanks! That last test passes on my machine the last time I checked, but this will also make it easy to proof everyone's dev environment. The one failed test aside, does this seem like a suitable test framework (even more realistic looking with one test failing)? I looked at some of the other test harnesses, and pytest seemed to have the most pleasing way to run a whole big test suite with nothing more than pytest -v and get nice output like this. But I have minimal experience.

**Comment by amykglen** [comment]
It seems good to me! (Though I also don't have much experience.)

Strangely, when I run `pytest -v`, it doesn't work for me (errors out complaining about `No module named 'cachecontrol'`, but it does work if I run `python -m pytest` (output looks like what @dkoslicki posted above).

**Comment by edeutsch** [comment]
If you do `which pytest` and `which python` or `which python3` are they in the same place?

**Comment by edeutsch** [comment]
So I found the source of the failing test. Someone had sneakily renamed overlay parameter virtual_edge_type to virtual_relation_label. I updated ARAX_query_graph_interpreter_templates.yaml, which holds the templates for DSL generation and all seems to be well again. Try a git pull and pytest again.

**Comment by dkoslicki** [comment]
> errors out complaining about No module named 'cachecontrol'

@amykglen do you have `CacheControl` installed? It’s listed in the [requirements.txt](https://github.com/RTXteam/RTX/blob/master/requirements.txt#L25)...

**Comment by dkoslicki** [comment]
> Someone had sneakily renamed overlay parameter virtual_edge_type to virtual_relation_label.

Ah yup, this was a result of the changes that we made for #786 (since the user is to specify the relation, not the type for virtual edges). I’ll pull and test again soon.

**Comment by amykglen** [comment]
Thanks, yep - I did have cachecontrol installed and `which python`/`which pytest` showed they were indeed in the same place.. but deactivating and reactivating my virtual environment seems to have done the trick for whatever reason - now `pytest -v` works just fine.

**Comment by dkoslicki** [comment]
Note: I've added test code for each existing `overlay()` command, as well as put a couple of more complicated workflows in `test_ARAX_workflows.py`.
@finnagin you may consider putting simple test cases for your filtering in `test/test_ARAX_filter_kg.py` and `test/test_ARAX_filter_results.py`. See [test_ARAX_overlay.py](https://github.com/RTXteam/RTX/blob/issue720/code/ARAX/test/test_ARAX_overlay.py) for how I did it.

**Comment by dkoslicki** [comment]
Oh, and @amykglen , you might consider moving the (relevant parts of) `ARAX_resultify_testcases.py` over to the `test` folder so we can use `pytest` on them.

**Comment by dkoslicki** [comment]
> Oh, and @amykglen , you might consider moving the (relevant parts of) `ARAX_resultify_testcases.py` over to the `test` folder so we can use `pytest` on them.

Never mind @amykglen , looks like it was a simple copy paste (and a couple small fixes), so it's now located at [test_ARAX_resultify.py](https://github.com/RTXteam/RTX/blob/issue720/code/ARAX/test/test_ARAX_resultify.py) and all tests are passing.

**Comment by amykglen** [comment]
oh nice, thanks! I'll get a `test_ARAX_expand.py` in there soon.

**Comment by saramsey** [comment]
@finnagin indicates that there are several test cases for ARAX_filter.py to be implemented; slated for implementation after the article submission.

**Comment by saramsey** [comment]
finn reports that he has several additional test cases in development, including updating for default behavior of filter

**Comment by edeutsch** [comment]
I haven't run tests in a while, so last night after finishing some development, I decided to run the whole test suite. It took nearly 15 hours to complete. This seems a bit undesirable.

I wonder if maybe we should have a test/ and a test_slow/ area? Where test/ could provide basic testing where each test should take only a few seconds max. Therefore it would be easy/fast for someone to run an entire test suite after some development without taking hours. And then test_slow could be reserved for more exhaustive testing for things that may take substantial amounts of time. Those tests could be run less often.

Or is there something wrong with my setup that it takes 15 hours? Maybe now that requests-caches are filled, it will be faster?

How long does it take for y'all? 
What do you think?

**Comment by amykglen** [comment]
yeah, the whole suite takes forever for me right now too, but I think the bulk of that slowdown is due to the temporary COHD issues (soon to be resolved by #875 I believe?) (if you skip tests involving COHD, I think the whole suite completes rather quickly..)

I agree it'd be nice to try to have fast tests though, and that it might be nice to have a way to skip slow tests when desired. it's possible to do that within the current pytest setup, if wanted - for example, if I add a little `conftest.py` file to `/test` to hook things up, then if you were to mark a test with `@pytest.mark.slow` above its `def` line, it would be skipped when `pytest -v` is run, but included when `pytest -v --runslow` is run.

**Comment by edeutsch** [comment]
I like that idea. I think it would be nice to be able to run the full suite of (reasonably fast) tests in like 5 minutes. Otherwise, they are less likely to be run. Then occasionally we could run the full slow suite.

**Comment by dkoslicki** [comment]
Major +1 to the conftest.py suggestion by @amykglen. I've had to resort to picking and choosing individual tests to run due to how long the COHD tests take.

**Comment by edeutsch** [comment]
My only concern is I wonder if may be easier to lose track of which are the slow ones if they are hidden with `@pytest.mark.slow`. If you run fast then you wouldn't see them. If you run --runslow then they're intermingled with the fast ones. One possible benefit of organizing them in a different directory is that all you could just run the slow ones and see them individually. I'm wondering if just physically separating them is easier. I dunno. whatever someone if willing to do.

**Comment by amykglen** [comment]
well, it took very little effort so I just pushed the `conftest.py` so we can try the `@pytest.mark.slow` method if desired. it still shows the skipped slow tests when you use `pytest -v`, like:
```
test_ARAX_expand.py::test_kg1_parkinsons_demo_example SKIPPED
test_ARAX_expand.py::test_kg2_parkinsons_demo_example PASSED  
```
but it's also easy to adjust the command line options to whatever we want - so I added a `--runonlyslow` option as well, if that's any consolation. :) (the downside I see to the separate directory method is that they couldn't all be run at once.)

**Comment by amykglen** [comment]
I updated the test [README](https://github.com/RTXteam/RTX/blob/master/code/ARAX/test/README.md) with these different options, fyi.

one other random thing pytest has that I've found handy, in case anyone isn't aware, is that if you use `--durations=10`, then when pytest is done running it'll show you the ten (or whatever number) slowest tests and their durations... (e.g., `pytest -v test_ARAX_resultify.py --durations=5`)

**Comment by dkoslicki** [comment]
Ok, overlay tests complete in about 1 min now.
Workflow tests are basically all marked as slow (since they are) save for the only fast one `test_example_3`

Leaving it up to others to address their tests and mark slow as needed.

Currently all tests (excluding slow) takes ~12min on my machine.

Note @edeutsch after particularly large architectural tests, it might be advisable to at least run all the `test_ARAX_worksflows.py` tests (even the slow ones) as while they may take a while, they are end-to-end tests for QGs that we've advertised as being able to answer.

**Comment by edeutsch** [comment]
great, thanks. git pulled and retesting locally.

**Comment by dkoslicki** [comment]
Ok, overlay down to 38 seconds

**Comment by amykglen** [comment]
went through the `expand` tests - down to 45 seconds when skipping 'slow' tests.

I'll go through `resultify`'s tests as well.

**Comment by edeutsch** [comment]
Great! quick test suite is passing for me. Takes 10 minutes total. test_example_3 is by far the slowest now. Maybe that can go into the slow boat?
```
============================================================================ slowest 10 test durations =============================================================================
234.32s call     test_ARAX_workflows.py::test_example_3
49.24s call     test_ARAX_filter_kg.py::test_default_std_dev
44.92s call     test_ARAX_resultify.py::test_issue731
40.18s call     test_ARAX_resultify.py::test_issue731b
33.69s call     test_ARAX_resultify.py::test_parallel_edges_between_nodes
23.93s call     test_ARAX_filter_kg.py::test_default_std_top_n
20.21s call     test_ARAX_resultify.py::test_issue680
16.68s call     test_ARAX_resultify.py::test_issue833
11.83s call     test_ARAX_resultify.py::test08
10.45s call     test_ARAX_query.py::test_query_by_query_graph_2
============================================================= 72 passed, 37 skipped, 11 warnings in 619.95s (0:10:19) ==============================================================
```

**Comment by amykglen** [comment]
alright, `resultify` is now down to about 20 seconds for fast tests, fyi

**Comment by amykglen** [comment]
re: @edeutsch's comment about `test_example_3` taking a long time - not sure if there's caching at play or the test changed since you ran it, but it's only taking 30 seconds for me:
```
============================================== slowest 10 test durations ===============================================
69.88s call     test_ARAX_filter_kg.py::test_default_std_dev
34.60s call     test_ARAX_workflows.py::test_example_3
34.19s call     test_ARAX_filter_kg.py::test_default_std_top_n
11.40s call     test_ARAX_overlay.py::test_FET_ex1
10.60s call     test_ARAX_overlay.py::test_predict_drug_treats_disease_virtual
10.05s call     test_ARAX_overlay.py::test_predict_drug_treats_disease_attribute
10.00s call     test_ARAX_overlay.py::test_issue_832
9.65s call     test_ARAX_overlay.py::test_issue_832_non_drug
7.85s call     test_ARAX_overlay.py::test_jaccard
7.15s call     test_ARAX_query.py::test_query_by_query_graph_2
=============================== 68 passed, 41 skipped, 10 warnings in 275.28s (0:04:35) ================================
```

**Comment by edeutsch** [comment]
You're right, must have been a caching issue or something else. It is now no longer the slowest test.

**Comment by dkoslicki** [comment]
@finnagin any idea why these tests are now taking a long time?
```
78.99s call     test_ARAX_filter_kg.py::test_default_std_dev
24.64s call     test_ARAX_filter_kg.py::test_default_std_top_n
24.03s call     test_ARAX_filter_kg.py::test_warning
```
I tried changing out the DOID:1588 for DOID:9008 since it's connected to less chemical substances, but that caused the tests to fail (since I think everything gets filtered out)

**Comment by edeutsch** [comment]
On arax.rtxi.ai, they are taking *way* longer than that:
```
1090.81s call     test_ARAX_filter_kg.py::test_default_std_dev
530.58s call     test_ARAX_filter_kg.py::test_default_std_top_n
```

**Comment by finnagin** [comment]
Woah, that is quite a bit slower than they used to take. Not sure why this is. I'll look into this.

**Comment by finnagin** [comment]
So I was able to fix this by overlaying normalized google distance instead for the prediction probabilities but I'm still not fully sure why it was taking so long. Will need to poke around later when it is not so late since it takes so long to run.

**Comment by edeutsch** [comment]
test in `master` is failing now for me:
```
FAILED test_ARAX_filter_kg.py::test_default_std_dev - assert 55 == 0
```

**Comment by finnagin** [comment]
That's odd it just worked when I tried it on my end. Might have been a hiccup when pinging ncbi eutils?

**Comment by edeutsch** [comment]
sorry for the false alarm, upon rerunning, it just worked. dunno what happened.

**Comment by finnagin** [comment]
No worries! Probably something with eutils. It's always a little spotty.

**Comment by dkoslicki** [comment]
@finnagin you may consider switching out the overlay with something that doesn’t relay on an API call (so we don’t get randomly failing tests due to someone else’s problem). Eg. Jaccard. Also note that the drug-disease prediction is relatively fast if the underlying KG is relatively small, so you could try my original suggestion of switching out the droid and adjusting the assertions

**Comment by edeutsch** [comment]
just ran the test again this morning for fun and again:
```
FAILED test_ARAX_filter_kg.py::test_default_std_dev - assert 55 == 0
```
so it seems unstable.

**Comment by edeutsch** [comment]
PASSED on the second run

**Comment by edeutsch** [comment]
pulled the latest `master` and test_example_3 is now failing for me:
First pass:
>       assert len(message.results) in range(45, 60)  # FIXME: BUG: sometimes the workflow returns 47 results, sometimes 48, sometime smore or less?! Probably due to API's not returning results (API calls to NGD and/or COHD sometimes don't go through)
E       AssertionError: assert 71 in range(45, 60)
393.85s call     test_ARAX_workflows.py::test_example_3

Second pass:
>       assert len(message.results) in range(45, 60)  # FIXME: BUG: sometimes the workflow returns 47 results, sometimes 48, sometime smore or less?! Probably due to API's not returning results (API calls to NGD and/or COHD sometimes don't go through)
E       AssertionError: assert 74 in range(45, 60)
109.68s call     test_ARAX_workflows.py::test_example_3

seems like repeated runs generates different numbers of results. More than expected.
Ideas?

**Comment by dkoslicki** [comment]
Same on my end. I think it’s due to changes in the filtering that @finnagin implemented, so he may want to have a look at this test case.

Thanks,

~David

**Comment by finnagin** [comment]
This might be due to the implementation of #846 since nodes might have been deleted erroneously before (if they shared the same curie id but different qnode_id) leading to a smaller set of results previously. I'll look into it.

**Comment by finnagin** [comment]
I did some poking around and wasn't able to find anything that filter was doing incorrectly. Looking at the results before and after the 2 filter calls it removes all the edges with attributes outside of the range specified. Also it can't be a side effect of #846 like I thought it might be earlier since there the query graph is (DOID:9406)-(chemical_substance)-(protein) and I don't believe there is any overlap of curie ids between disease, chemical substance or protein nodes. Is it possible the results we are getting from COHD or NCBI are different? 

If I change the code on line 133 from range(45, 60) to range(65, 80) to get it to pass the result count check it now fails for me at the `__virtual_tester` check for the 'observed_expected_ratio' edges and it looks like this is because it checks for at least 2 unique values of the type 'observed_expected_ratio' however there is only one. if I comment out the code from filter on i.e.:
```
    query = {"previous_message_processing_plan": {"processing_actions": [
        "add_qnode(name=DOID:9406, id=n00)",
        "add_qnode(type=chemical_substance, is_set=true, id=n01)",
        "add_qnode(type=protein, id=n02)",
        "add_qedge(source_id=n00, target_id=n01, id=e00)",
        "add_qedge(source_id=n01, target_id=n02, id=e01)",
        "expand(edge_id=[e00,e01])",
        "overlay(action=overlay_clinical_info, observed_expected_ratio=true, virtual_relation_label=C1, source_qnode_id=n00, target_qnode_id=n01)",
        #"filter_kg(action=remove_edges_by_attribute, edge_attribute=observed_expected_ratio, direction=below, threshold=3, remove_connected_nodes=t, qnode_id=n01)",
        #"filter_kg(action=remove_orphaned_nodes, node_type=protein)",
        #"overlay(action=compute_ngd, virtual_relation_label=N1, source_qnode_id=n01, target_qnode_id=n02)",
        #"filter_kg(action=remove_edges_by_attribute, edge_attribute=normalized_google_distance, direction=above, threshold=0.85, remove_connected_nodes=t, qnode_id=n02)",
        "resultify(ignore_edge_direction=true, debug=true)",
        "return(message=true, store=false)"
    ]}}
```
and then I sort and print the unique 'observed_expected_ratio' values and I get:
```
[-inf, -0.06422247, -0.0377248, 0.001999439, 0.4254316, 0.5473396, 0.5945498, 0.6448684, 0.7068931, 0.8239506, 0.9234182, 0.9607711, 1.214741, 3.803441]
```
Since we are filtering all values below 3 it removes every value listed above except 3.803441 and thus fails the `__virtual_tester` check as there is only 1 unique value left.

Could this be related? This makes me think we might be getting different results from COHD as we're testing if there are at least 2 edges with unique values larger than 3 but now we only are only getting 1.

**Comment by edeutsch** [comment]
This is still a problem. I have not gotten a clean test run in weeks now. only one error now equals success. Not ideal, but okay. Anyone able to repro or fix?

**Comment by edeutsch** [comment]
test_ARAX_workflows.py::test_example_3 FAILED                                                                                                       [ 90%]
test_ARAX_workflows.py:133: AssertionError
E       AssertionError: assert 72 in range(45, 60)

**Comment by finnagin** [comment]
So I think that particular error can be fixed by changing range(45, 60) to range(65, 80) as we are getting more results consistently now. 

However, I think there is something else going on that we should probably look into as before we start changing numbers until it works since a second error appears relating to the 'observed_expected_ratio' virtual edge. As I mentioned in my last post it looks to me like this is potentially something caused by the clinical info overlay step as we are getting only 1 unique value above 3 now. @dkoslicki  or @chunyuma do you know why there might be only 1 unique 'observed_expected_ratio' value now?

**Comment by edeutsch** [comment]
well, that seemed to quiet things for a while, but now I'm seeing this in the latest test:
E       AssertionError: assert 82 in range(60, 80)

@finnagin would you address this so that we can have all tests passing?

**Comment by finnagin** [comment]
Sure @edeutsch ! I've changed the range so that it should pass now with that many results.

**Comment by dkoslicki** [comment]
This has been in place for a while and appears to be working. Feel free to object if you feel otherwise
